server/src/main/java/keywhiz/service/resources/BatchSecretDeliveryResource.java

```diff
@@ -0,0 +1,156 @@
+    this.secretController = secretController;
+    this.aclDAO = aclDAOFactory.readonly();
+    this.clientDAO = clientDAOFactory.readwrite();
+  }
+
+  @VisibleForTesting
+  BatchSecretDeliveryResource(SecretController secretController, AclDAO aclDAO,
+                              ClientDAO clientDAO) {
+    this.secretController = secretController;
+    this.aclDAO = aclDAO;
+    this.clientDAO = clientDAO;
+  }
+
+  /**
+   * Retrieve Secret by name
+   *
+   * @param secrets the name of the Secrets to retrieve in batch
+   * @param client  the client performing the retrieval
+   * @return the secret with the specified name, if present and accessible to the client
+   * <p>
+   * responseMessage 200 Found and retrieved Secret with given name
+   * responseMessage 403 Secret is not assigned to Client
+   * responseMessage 404 Secret with given name not found
+   * responseMessage 500 Secret response could not be generated for given Secret
+   */
+  @Timed
+  @ExceptionMetered
+  @POST
+  @Consumes(APPLICATION_JSON)
+  public List<SecretDeliveryResponse> getBatchSecret(@Auth Client client, @Valid BatchSecretRequest request) {
+
+
+    List<SanitizedSecret> sanitizedSecrets = aclDAO.getBatchSanitizedSecretsFor(client, request.secrets);
+    List<Secret> secrets = secretController.getSecretsByName(request.secrets);
+
+    boolean clientExists = clientDAO.getClient(client.getName()).isPresent();
+
+
+    Iterator reqsecretsiter = request.secrets.iterator();
+
+
+    // The request fails whenever a single secret is requested that is not accessible
+    // The client is responsible for only requesting secrets they have permission for
+    while (reqsecretsiter.hasNext()) {
+      String secretname = (String) reqsecretsiter.next();
+
+      // We iterate over both arrays since the order of the two arrays might not match
+      SanitizedSecret sanitizedsecret = null;
+      for (SanitizedSecret s : sanitizedSecrets) {
+        if (s.name().equals(secretname)) {
+          sanitizedsecret = s;
+          break;
+        }
+      }
+
+      Secret secret = null;
+      for (Secret s : secrets) {
+        if (s.getName().equals(secretname)) {
+          secret = s;
+          break;
+        }
+      }
+
+      boolean secretExists = secret != null;
+      boolean sanitzedsecretExists = sanitizedsecret != null;
+
+
+      if (!sanitzedsecretExists) {
+        if (clientExists && secretExists) {
+          throw new ForbiddenException(format("Access denied: %s at '%s' by '%s'", client.getName(),
+                  "/batchsecret " + secretname, client));
+        } else {
+          if (clientExists) {
+            logger.info("Client {} requested unknown secret {}", client.getName(), secretname);
+          }
+          throw new NotFoundException();
+
+        }
+      }
+    }
+
+    logger.info("Client {} granted access to {}.", client.getName(), sanitizedSecrets);
```

Comment:
nit: maybe just log the secret names here, rather than the full SanitizedSecret data which includes all fields except the secret contents, for more readable log messages.

`sanitizedSecrets.stream().map(s -> s.getName()).collect(toList())` should do it